libs/ml/examples/distributed_learning/main.cpp

```diff
@@ -0,0 +1,420 @@
+
+    // Clear loss file
+    std::ofstream lossfile("losses_" + id_ + ".csv", std::ofstream::out | std::ofstream::trunc);
+    lossfile.close();
+  }
+
+  void SetCoordinator(std::shared_ptr<Coordinator> &coordinator_ptr)
+  {
+    coordinator_ptr_ = coordinator_ptr;
+  }
+
+  /**
+   * Main loop that runs in thread
+   */
+  void MainLoop()
+  {
+    std::ofstream lossfile("losses_" + id_ + ".csv", std::ofstream::out | std::ofstream::app);
+    DataType      loss;
+
+    while (coordinator_ptr_->state == CoordinatorState::RUN)
+    {
+      // Shuffle the peers list to get new contact for next update
+      fetch::random::Shuffle(gen_, peers_, peers_);
+
+      // Train one batch to create own gradient
+      Train();
+
+      // Put own gradient to peers queues
+      BroadcastGradients();
+
+      // Load own gradient
+      TensorVectorType gradients = g_.GetGradients();
+      TensorVectorType new_gradients;
+
+      // Sum all gradient in queue
+      while (!gradient_queue.empty())
+      {
+        {
+          std::lock_guard<std::mutex> l(queue_mutex_);
+          new_gradients = gradient_queue.front();
+          gradient_queue.pop();
+        }
+
+        for (SizeType j{0}; j < gradients.size(); j++)
+        {
+          fetch::math::Add(gradients.at(j), new_gradients.at(j), gradients.at(j));
+        }
+      }
+
+      // Apply sum of all gradients from queue along with own gradient
+      ApplyGradient(gradients);
+
+      // Validate loss for logging purpose
+      Test(loss);
+
+      // Save loss variation data
+      // Upload to https://plot.ly/create/#/ for visualisation
+      if (lossfile)
+      {
+        lossfile << GetTimeStamp() << ", " << loss << "\n";
+      }
+    }
+
+    lossfile << GetTimeStamp() << ", "
+             << "STOPPED"
+             << "\n";
+    lossfile.close();
+  }
+
+  /**
+   * Train one batch
+   * @return training batch loss
+   */
+  DataType Train()
+  {
+    dataloader_.SetMode(fetch::ml::dataloaders::DataLoaderMode::TRAIN);
+    dataloader_.SetRandomMode(true);
+
+    DataType loss        = 0;
+    bool     is_done_set = false;
+
+    std::pair<TensorType, std::vector<TensorType>> input;
+    input = dataloader_.PrepareBatch(batch_size_, is_done_set);
+
+    {
+      std::lock_guard<std::mutex> l(model_mutex_);
+
+      g_.SetInput("Input", input.second.at(0));
+      g_.SetInput("Label", input.first);
+
+      TensorType loss_tensor = g_.ForwardPropagate("Error");
+      loss                   = *(loss_tensor.begin());
+      g_.BackPropagateError("Error");
+    }
+
+    return loss;
+  }
+
+  /**
+   * Run model on test set to get test loss
+   * @param test_loss
+   */
+  void Test(DataType &test_loss)
+  {
+    dataloader_.SetMode(fetch::ml::dataloaders::DataLoaderMode::TEST);
+
+    // Disable random to run model on whole test set
+    dataloader_.SetRandomMode(false);
+
+    SizeType test_set_size = dataloader_.Size();
+
+    dataloader_.Reset();
+    bool is_done_set;
```

Comment:
presumably you should give a default initialisation to is_done_set (i.e. false?)